### ISSUE TITLE
[Prometheus and OpenMetrics] Handle scheme URL and scope attributes as identifying

### ISSUE THREAD

**Issue by pellared** [open-telemetry/opentelemetry-specification#4223]
I think this needs updates to some of the compatibility docs (at least [prometheus](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#instrumentation-scope)).  Currently, Prometheus adds scope name and scope version as labels to ensure we don't end up with duplicate timeseries, and we are able to put scope attributes in a separate `otel_scope_info` metric. With this change, we will need to update this to instead add the scope name, scope version, and scope attributes to all metrics.

TIL `schema_url` is also identifying, so that is something we will need to handle in Prometheus export and other "non-OTLP" compatibility documents.

_Originally posted by @dashpole in https://github.com/open-telemetry/opentelemetry-specification/issues/4161#issuecomment-2361470786_

**Comment by pellared** [comment]
@dashpole, are you able to take this?

**Comment by dashpole** [comment]
cc @open-telemetry/wg-prometheus 

Yes, I'll take this

**Comment by pellared** [comment]
From https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#instrumentation-scope-1:

> Scope attributes MUST also be added as labels following the rules described in the [Metric Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#metric-attributes) section below.

I feel that we only should add `otel__scheme_url` which is also identifying. All of the rest is already defined.

Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#get-a-meter

CC @dashpole

**Comment by pellared** [comment]
The idea to add `otel_scope_[attribute]` and `otel_scheme_url` labels to metrics.

**Comment by jsuereth** [comment]
This seems like the right path going forward.  Now that these are all identifying, we CANNOT rely on joins for `scope`.

**Comment by pellared** [comment]
@open-telemetry/prometheus-interoperability, @open-telemetry/collector-maintainers: PTAL

**Comment by jade-guiton-dd** [comment]
On the Collector side, I think there are no issues as long as the scope attributes are 1. exposed to the user, and 2. identifying on the Prometheus side, to avoid errors with time series collisions. Adding them as `otel_scope_[attribute]` labels on the metric points themselves as proposed above sounds like the simplest way to accomplish this.

(Unrelated, but what is the distinction between sub-issues 1 and 2 above?)

**Comment by pellared** [comment]
> (Unrelated, but what is the distinction between sub-issues 1 and 2 above?)

@jade-guiton-dd 

I have forgotten to delete "and scope attributes"` from sub-issue 1 😅  Fixed 😉 

I will do my best to move this forward. However, I will start in about 2 weeks. 

In the meantime you can try prototype  it on the collector-side and to test everything e2e with https://github.com/open-telemetry/opentelemetry-go/pull/5947 to make sure that there are not issues with this proposal. Having such e2e prototypes would help move the specification work forward.

**Comment by aknuds1** [comment]
I made a corresponding [Prometheus issue](https://github.com/prometheus/prometheus/issues/16557).

**Comment by jade-guiton-dd** [comment]
@pellared By "prototyping on the Collector side", did you mean in the Prometheus exporter in collector-contrib? Based on the discussion in today's Collector SIG meeting, it sounds like [this part in CONTRIBUTING.md](https://github.com/open-telemetry/opentelemetry-specification/blob/cdac03154c5b814a83689aecb97f82d6bbf7c0b2/CONTRIBUTING.md?plain=1#L39) would make this not "count" as a prototype for the purpose of changing the spec:

> For this reason, when adding new features with Development maturity level, a prototype is defined as a working demonstration of the feature in a language implementation which has the support of that language's maintainers.

In other words, we may need a prototype in another language SDK's Prometheus exporter

**Comment by mx-psi** [comment]
Filed #4497 to reword the prototype requirement

**Comment by pellared** [comment]
@jade-guiton-dd, I think that Prometheus receiver in collector-contrib is also crucial.

**Comment by jade-guiton-dd** [comment]
Yes of course, everything performing translation between OTel and Prometheus will need to be updated eventually. I was just wondering about what prototypes you think need to be done to get the spec change moving.

Re-reading your message, I may have misunderstood. By "Collector + Go", did you mean a single prototype where the draft PR modifying the SDK exporter is integrated into the Collector to test that it fixes our issue when producing scope attributes?

**Comment by pellared** [comment]
> did you mean a single prototype where the draft PR modifying the SDK exporter is integrated into the Collector to test that it fixes our issue when producing scope attributes?

No. I just mean building 3 prototypes: Go SDK Prometheus exporter, Collector Prometheus receiver, Collector Prometheus exporter and check if they will work together correctly (even "manually"). I just think it would be nice to check all components starting from an instrumented application, through Collector, finishing in Prometheus. I am fine with any other validation/testing approach.

**Comment by pellared** [comment]
I updated the OTel Go's exporter implementation here: https://github.com/open-telemetry/opentelemetry-go/pull/5947. PTAL

**Comment by ArthurSens** [comment]
PoC for Prometheus exporter in collector-contrib: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40004

Reviews appreciated, especially in test coverage of edge cases :)

**Comment by pellared** [comment]
@ArthurSens, are you going to also work on prototype for [prometheusreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver)?

**Comment by pellared** [comment]
Given https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40004#issue-3057707170:

> Turns out our exporter never exposed otel_scope_info, so this PR just introduces the addition of the new labels.

and https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/3c6a5ea6430afaac7c33a25ff276903ef829693b/receiver/prometheusreceiver#resource-and-scope:

> It drops the `otel_scope_info` metric, and uses attributes (other than `otel_scope_name` and `otel_scope_version`) to populate Scope Attributes.

It looks like we can drop point 3 from https://github.com/open-telemetry/opentelemetry-specification/issues/4223#issuecomment-2839595942 completely (`otel_scope_info` metric is useless)

**Comment by ArthurSens** [comment]
I spent some time yesterday and it was not as easy as I expected, but I'm also not 100% familiar with that codebase 😬

I plan to spend more time on it this week, but it might take a while to have a PoC

**Comment by pellared** [comment]
I created a specification PR. PTAL:
- https://github.com/open-telemetry/opentelemetry-specification/pull/4505

**Comment by ArthurSens** [comment]
PoCs in collector are ready for review:

* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40060
* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40004
